src/coreclr/jit/optimizer.cpp

```diff
@@ -2904,30 +2925,119 @@ bool Compiler::optCanonicalizeLoopNest(unsigned char loopInd)
 }
 
 //-----------------------------------------------------------------------------
-// optCanonicalizeLoop: ensure that each loop top's back edges do not come
-//   from nested loops.
+// optCanonicalizeLoop: ensure that each loop top's back edges come only from
+//   blocks in the same loop.
 //
 // Arguments:
 //   loopInd - index of the loop to consider
 //
 // Returns:
 //   true if flow changes were made
 //
+// Notes:
+//
+// Back edges incident on loop top fall into one three groups:
+// 
+// (1) Outer non-loop backedges (preds dominated by enter where pred is not in loop)
+// (2) The canonical backedge (pred == bottom)
+// (3) Nested loop backedges or nested non-loop backedges 
+//     (preds dominated by enter, where pred is in loop, pred != bottom)
+//
+// We assume dominance has already been established by loop recognition (that is,
+// anything classfied as a loop will have all backedges dominated by loop entry,
+// so the only possible non-backedge predecessor of top will be head).
+//
+// We cannot check dominance here as the flow graph is being modified.
+//
+// If either set (1) or (3) is non-empty the loop is not canonical.
+//
+// This method will split the loop top into two or three blocks depending on
+// whether (1) or (3) is non-empty, and redirect the edges accordingly.
+//
+// Loops are canoncalized outer to inner, so inner loops should never see outer loop 
+// non-backedges, as the parent loop canonicalization should have handled them.
+//
 bool Compiler::optCanonicalizeLoop(unsigned char loopInd)
 {
-    // Is the top uniquely part of the current loop?
+    bool modified = false;
+    BasicBlock* const b = optLoopTable[loopInd].lpBottom;
     BasicBlock* const t = optLoopTable[loopInd].lpTop;
+    BasicBlock* const h = optLoopTable[loopInd].lpHead;
+    BasicBlock* const e = optLoopTable[loopInd].lpEntry;
+
+    // Look for case (1)
+    //
+    bool doOuterCanon = false;
 
-    if (t->bbNatLoopNum == loopInd)
+    for (BasicBlock* const topPredBlock : t->PredBlocks())
     {
-        return false;
+        const bool predIsInLoop = (t->bbNum <= topPredBlock->bbNum) && (topPredBlock->bbNum <= b->bbNum);
+        if (predIsInLoop || (topPredBlock == h))
+        {
+            // no action needed
+        }
+        else
+        {
+            JITDUMP("in optCanonicalizeLoop: " FMT_LP " top " FMT_BB " (entry " FMT_BB " bottom " FMT_BB ") %shas a non-loop backedge from " FMT_BB 
+                "\n",
```

Comment:
```suggestion
                "%s\n",
```
missing replacement for "need to canonicalize non-loop backedges" output